File: solutions/java/queue/step03_dequeue/Queue.java
import java.util.ArrayList;
import java.util.List;

public class Queue<T> {
    private final List<T> items;
    private int head;

    public Queue() {
        this.items = new ArrayList<>();
        this.head = 0;
    }

    public void enqueue(T value) {
        items.add(value);
    }

    public T dequeue() {
        if (head >= items.size()) {
            return null;
        }
        T value = items.get(head);
        head += 1;
        return value;
    }
}

File: solutions/python/queue/03_dequeue/queue.py
class Queue:
    def __init__(self) -> None:
        self._items = []
        self._head = 0

    def enqueue(self, value) -> None:
        self._items.append(value)

    def dequeue(self):
        if self._head >= len(self._items):
            return None
        value = self._items[self._head]
        self._head += 1
        return value

